pkg/maps/policymap/policymap.go

```diff
@@ -162,6 +162,31 @@ func (pm *PolicyMap) DumpToSlice() ([]PolicyEntryDump, error) {
 	return entries, nil
 }
 
+// Flush deletes all entries from the given policy map
+func (pm *PolicyMap) Flush() error {
+	var key, nextKey policyKey
+	for {
+		err := bpf.GetNextKey(
+			pm.Fd,
+			unsafe.Pointer(&key),
+			unsafe.Pointer(&nextKey),
+		)
+
+		// FIXME: Ignore delete errors?
+		bpf.DeleteElement(
+			pm.Fd,
+			unsafe.Pointer(&key),
+		)
+
+		if err != nil {
```

Comment:
Yes, because get an error saying `nextKey` is not available but we can still delete the current `key`. @borkmann do we have a way to check the error message? if error is because of `nextKey` or `key`?

Reply:
Currently, `GetNextKey` is a pretty thin wrapper around the syscall with the behaviour described here:
http://man7.org/linux/man-pages/man2/bpf.2.html

If we wanted to avoid delete errors, we should allow `GetNextKey` to return whether the error is ENOENT (ie, `key` is valid but it's the last valid key). Then we could report all (other) errors returned by `GetNextKey`.

Reply:
Hmm, we also suppress reporting all the errors because we `return nil` below. Making `GetNextKey` return a bool (or something) to indicate last would also be a good thing. Other options include setting nextKey to null and checking that but the more explicit return flag is more obvious.

Reply:
I think the most reasonable thing would imo be: add a new API to our BPF golang interface that can be used to retrieve the first key in the map. Newer kernels have the behaviour that when key is NULL, we search and return the first key in the map. Older kernels don't have this and only get you the first key when you have a lookup mismatch, usually with an all 0 bytes key, but it's a bit fragile. But we should have the API such that we try first the NULL key and then fallback to the 0 bytes one. I agree with @joestringer that it might make sense to change the API to properly return errors. @raybejjani if flush fails, how should the daemon behave in that case? Exit ungracefully?

Reply:
I'll defer to aanm on whether we should crash out if flush fails (since he was adding this code). I also don't really know. I would normally say we should print clear errors and keep working as best as we can but this is a globally shared table so it's more important, and an error here might mean we can't do anything. If we did exit, we'd likely be restarted and repeat all this, except we have partially already tried the restore stuff. Also, would anything change if we restarted vs just kept going and tried to apply policy anyway.

Reply:
The exact behavior here is part of a bigger discussion on what to do when policy update fails. We need to disconnect the endpoint and put it into error state as we can't guarantee the security behavior. This should not be part of this PR though.